File: coodingpenguin/leetcode/139_WordBreak.py
# 문제: 139. Word Break
# 링크: https://leetcode.com/problems/word-break/

# 시간/공간: 32ms / 14.3MB
# 참고: https://leetcode.com/problems/word-break/discuss/43808/Simple-DP-solution-in-Python-with-description
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        table = [False] * (len(s) + 1)
        table[0] = True
        for i in range(1, len(s) + 1):
            for word in wordDict:
                if table[i - len(word)] and s[i - len(word) : i] == word:
                    table[i] = True
        return table[-1]

File: sally/leetcode/139_Word Break.cpp
// 139. Word Break
// Runtime: 40 ms, faster than 15.17 % of C++ online submissions for Word Break.
// Memory Usage : 13.1 MB, less than 51.69 % of C++ online submissions for Word Break.
// ref: https://leetcode.com/problems/word-break/discuss/43814/C%2B%2B-Dynamic-Programming-simple-and-fast-solution-(4ms)-with-optimization

class Solution {
public:
    bool dp[302] = { false, };
    bool wordBreak(string s, vector<string>& words) {
        int s_len = s.size();
        dp[0] = true;
        for (int i = 1; i <= s_len; i++) {
            for (int j = 0; j < i; j++) {
                if (dp[j]) {
                    string target = s.substr(j, i - j);
                    for (auto k : words)
                        if (k == target) {
                            dp[i] = true;
                        }
                }
            }
        }
        for (auto i : dp) cout << i << ' ';
        cout << '\n';
        return dp[s_len];
    }
};